plugins/module_utils/vm_snapshot.py

```diff
@@ -160,5 +180,38 @@ def filter_snapshots_by_params(
             new_snaps = [vm_snap for vm_snap in new_snaps if vm_snap["vm"]["snapshot_serial_number"] == params["serial"]]  # type: ignore
         if params["label"]:
             new_snaps = [vm_snap for vm_snap in new_snaps if vm_snap["label"] == params["label"]]  # type: ignore
-
+        if params["source_disk_type"]:
+            new_snaps = [(vm_snap, device_snap) for vm_snap in new_snaps for device_snap in vm_snap["device_snapshots"] if device_snap["type"] == params["source_disk_type"]]  # type: ignore
+        if params["source_disk_slot"]:
+            new_snaps = [(vm_snap, device_snap) for vm_snap in new_snaps for device_snap in vm_snap["device_snapshots"] if device_snap["slot"] == params["source_disk_slot"]]  # type: ignore
+        if params["source_disk_uuid"]:
+            new_snaps = [(vm_snap, device_snap) for vm_snap in new_snaps for device_snap in vm_snap["device_snapshots"] if device_snap["uuid"] == params["source_disk_uuid"]]  # type: ignore
         return new_snaps
+
+    @classmethod
+    def get_source_disk_info(
+        cls,
+        device_snapshots: List[Dict[Any, Any]],
+        disk_uuid: str,
+    ) -> Optional[Dict[Any, Any]]:
+        for device_snapshot in device_snapshots:
+            if device_snapshot["uuid"] == disk_uuid:
+                return device_snapshot
+        return None
+
+    @classmethod
+    def get_device_snapshot(
+        cls,
+        vm_snapshot = TypedVMSnapshotToAnsible,
+        filter: Dict[Any, Any]
+    ) -> Optional[Dict[Any, Any]]:
+        new_device_snaps = vm_snapshot.device_snapshots[:]
+
+        if filter["uuid"]:
+            new_device_snaps = [device_snap for device_snap in cls.device_snapshots if device_snap["uuid"] == filter["uuid"]]
+        if filter["slot"]:
+            new_device_snaps = [device_snap for device_snap in cls.device_snapshots if device_snap["slot"] == filter["slot"]]
+        if filter["type"]:
+            new_device_snaps = [device_snap for device_snap in cls.device_snapshots if device_snap["type"] == filter["type"]]
+
+        return new_device_snaps[0]
```

Comment:
What if there is more than 1 element in `new_device_snaps`? If you check for this case, and just raise ScaleComputingError (or subclass of it) - it will be safer.